ui/src/app/shared/header/header.component.ts

```diff
@@ -0,0 +1,148 @@
+import {ActivatedRoute, Router} from "@angular/router";
+import {JobStatus} from "../model/JobStatus";
+import {QueryJobsRequest} from "../model/QueryJobsRequest";
+import {environment} from "../../../environments/environment";
+import {QueryFields} from "../common";
+
+@Component({
+  selector: 'jm-header',
+  templateUrl: './header.component.html',
+  styleUrls: ['./header.component.css'],
+})
+export class HeaderComponent implements OnInit {
+  separatorKeysCodes = [ENTER];
+  control: FormControl = new FormControl();
+  options: string[] = [];
+  currentChipKey: string = "";
+  currentChipValue: string = "";
+  inputValue: string = "";
+
+  chips: Map<string, string>;
+
+  filteredOptions: Observable<string[]>;
+
+  constructor(
+    private readonly route: ActivatedRoute,
+    private readonly router: Router,
+    private zone: NgZone,
+  ) {
+    route.queryParams.subscribe(
+      params => this.refreshChips(params['q']))
+  }
+
+  ngOnInit(): void {
+    if (this.route.snapshot.queryParams['q']) {
+      this.chips = URLSearchParamsUtils.getChips(this.route.snapshot.queryParams['q']);
+    }
+    this.filterOptions();
+    this.options = URLSearchParamsUtils.getQueryFields();
+    if (environment.additionalColumns) {
+      this.options = this.options.concat(environment.additionalColumns);
+    }
+  }
+
+  addChip(value: string): void {
+    if ((value || '').trim()) {
+      if (value.split(': ').length == 2) {
+        // Parse as a full key:value pair
+        let keyVal: string[] = value.split(': ');
+        this.deleteChipIfExists(keyVal[0]);
+        this.chips.set(keyVal[0].trim(), keyVal[1].trim());
+      }
+      else {
+        // Parse as just the key
+        this.deleteChipIfExists(value);
+        this.chips.set(value.trim(), '');
+      }
+      this.inputValue = "";
+    }
+  }
+
+  assignChipValue(): void {
+    this.removeChip(this.currentChipKey);
+    this.chips.set(this.currentChipKey, this.currentChipValue);
+  }
+
+  filter(val: string): string[] {
+    return this.options.filter(option =>
+      option.toLowerCase().indexOf(val.toLowerCase()) === 0);
+  }
+
+  filterOptions(): void {
+    this.filteredOptions = this.control.valueChanges
+      .startWith(null)
+      .map(val => val ? this.filter(val) : this.options.slice());
+  }
+
+  getChipKeys(): string[] {
+    return Array.from(this.chips.keys());
+  }
+
+  getDisplayValue(chipKey: string) {
+    return chipKey + ': ' + this.chips.get(chipKey);
+  }
+
+  navigateWithStatus(statuses: JobStatus[]): void {
+    let query: QueryJobsRequest = URLSearchParamsUtils.unpackURLSearchParams(this.route.snapshot.queryParams['q']);
+    query.statuses = statuses;
+    this.router.navigate(
+      ['jobs'],
+      {queryParams: { q: URLSearchParamsUtils.encodeURLSearchParams(query)}}
+    );
+  }
+
+  removeChip(chipKey: string): void {
+    this.chips.delete(chipKey);
+  }
+
+  // TODO: Cut the dependency on string parsing to represent lists here
+  search(): void {
```

Comment:
Do we want to just automatically update after adding/removing a chip or is this the preferred behavior? I didn't think it was working until I looked at the code and saw we have to click search. Maybe we could make that button bigger?

Reply:
Sounds good, definitely don't do a lot of work to make the search button bigger if we might remove it at some point.